File: apps/gatekeeper/pkg/attestation/verify.go
package attestation

import (
	"context"
	"encoding/json"
	"strings"
	"time"

	"github.com/Super-Protocol/confidential-router/apps/gatekeeper/pkg/attestation/internal/certparse"
)

// AllowedClockSkew is how far into the future a bundle may be dated before the
// freshness check treats it as a violation rather than benign clock drift.
const AllowedClockSkew = 60 * time.Second

// Params configures a verification. Only Hostname and TrustedRoots are
// required; the zero value of every other field selects the default described
// on it.
type Params struct {
	// Hostname the bundle is published for. The bundle and the JWS payload must
	// both name it.
	Hostname string

	// TrustedRoots is the user's trust store. An empty store is legal and denies
	// everything at the untrusted-root stage.
	TrustedRoots []TrustedRoot

	// ObservedTLSFingerprint pins the channel binding. VerifyHostname fills it
	// in from its own handshake when left empty; VerifyBundle treats an empty
	// value as "no channel access" and falls back to the bundle's tlsLeaf.
	ObservedTLSFingerprint string

	// MaxBundleAge rejects bundles whose payload.issuedAt is older than this.
	// Zero disables the check, matching the TypeScript default.
	MaxBundleAge time.Duration

	// Now overrides the clock for the validity window and the freshness check.
	Now time.Time

	// Fetch tunes the evidence request made by VerifyHostname.
	Fetch FetchOptions

	// Fetcher overrides how the evidence document is retrieved. Nil uses Fetch,
	// which is the only implementation that can observe the TLS channel; a
	// substitute must supply ObservedTLSFingerprint itself or the verdict falls
	// back to the producer-asserted binding.
	Fetcher Fetcher
}

func (p Params) now() time.Time {
	if p.Now.IsZero() {
		return time.Now()
	}
	return p.Now
}

// VerifyHostname fetches the endpoint's evidence bundle and verifies it,
// binding the verdict to the TLS certificate observed on that same fetch.
//
// It never returns an error: every failure is a Result with the stage it
// happened at, so a caller can render or log the denial without unwrapping.
func VerifyHostname(ctx context.Context, params Params) Result {
	if params.Hostname == "" {
		return fail(StageFetch, "hostname must be a non-empty string")
	}
	if params.ObservedTLSFingerprint != "" && !IsFingerprint(params.ObservedTLSFingerprint) {
		return fail(StageTLSFingerprint, "observedTlsFingerprint must match sha256/<base64url>")
	}

	fetcher := params.Fetcher
	if fetcher == nil {
		fetcher = Fetch
	}
	fetched, err := fetcher(ctx, params.Hostname, params.Fetch)
	if err != nil {
		return fail(StageFetch, "%v", err)
	}
	if fetched.StatusCode < 200 || fetched.StatusCode >= 300 {
		result := fail(StageFetch, "unexpected status %d from %s", fetched.StatusCode, fetched.URL)
		result.ObservedTLSFingerprint = fetched.ObservedTLSFingerprint
		return result
	}

	// A caller may pin the fingerprint it expects — re-attestation against the
	// leaf a live verdict was formed over. The pin is checked against what this
	// exchange actually presented rather than replacing it, so a rotated (or
	// substituted) certificate is caught here instead of silently binding the
	// verdict to a certificate nobody saw.
	if params.ObservedTLSFingerprint == "" {
		params.ObservedTLSFingerprint = fetched.ObservedTLSFingerprint
	} else if !FingerprintsEqual(params.ObservedTLSFingerprint, fetched.ObservedTLSFingerprint) {
		result := fail(StageTLSFingerprint, "pinned observedTlsFingerprint %s is not the certificate served (%s)",
			params.ObservedTLSFingerprint, fetched.ObservedTLSFingerprint)
		result.ObservedTLSFingerprint = fetched.ObservedTLSFingerprint
		return result
	}

	result := VerifyBundle(fetched.Body, params)
	result.ObservedTLSFingerprint = fetched.ObservedTLSFingerprint
	return result
}

// VerifyBundle runs the verification pipeline over an already-retrieved bundle
// document. It is the offline entry point — `gatekeeper policy test --bundle`,
// the conformance fixtures — and the shared core of VerifyHostname.
//
// The stages, in order, are fetch (shape) → cert-chain → untrusted-root → jws →
// freshness → tls-fingerprint, matching ADR-003 §1 and the TypeScript verifier.
func VerifyBundle(document []byte, params Params) Result {
	if params.Hostname == "" {
		return fail(StageFetch, "hostname must be a non-empty string")
	}
	if params.ObservedTLSFingerprint != "" && !IsFingerprint(params.ObservedTLSFingerprint) {
		return fail(StageTLSFingerprint, "observedTlsFingerprint must match sha256/<base64url>")
	}

	bundle, errResult := parseBundle(document, params.Hostname)
	if bundle == nil {
		return errResult
	}

	chain, err := ValidateChain(bundle.CertChain, params.now())
	if err != nil {
		return fail(StageCertChain, "%v", err)
	}

	matchedRoot, errResult := matchTrustedRoot(chain.RootFingerprint, params.TrustedRoots)
	if matchedRoot == nil {
		return errResult
	}

	payload, errResult := verifyPayload(bundle, chain.leaf(), params.Hostname)
	if payload == nil {
		return errResult
	}

	if errResult, ok := checkFreshness(payload.Base().IssuedAt, params.MaxBundleAge, params.now()); !ok {
		return errResult
	}

	binding, errResult := checkChannelBinding(payload.Base().CertFingerprint, params.ObservedTLSFingerprint, bundle.TLSLeaf)
	if binding == "" {
		return errResult
	}

	return Result{
		OK:                     true,
		Kind:                   payload.Base().Kind,
		Payload:                payload,
		MatchedRoot:            *matchedRoot,
		RootCaTeeQuote:         bundle.RootCaTeeQuote,
		ChannelBinding:         binding,
		ObservedTLSFingerprint: params.ObservedTLSFingerprint,
	}
}

// parseBundle decodes the document and enforces the bundle shape, field by
// field in the order the TypeScript verifier checks them. Shape failures are
// reported at the fetch stage: from the caller's point of view the endpoint did
// not serve a usable bundle.
func parseBundle(document []byte, expectedHostname string) (*Bundle, Result) {
	var probe map[string]json.RawMessage
	if err := json.Unmarshal(document, &probe); err != nil {
		var anyValue any
		if json.Unmarshal(document, &anyValue) == nil {
			return nil, fail(StageFetch, "response body is not a JSON object")
		}
		return nil, fail(StageFetch, "failed to parse response body as JSON: %v", err)
	}
	if probe == nil {
		return nil, fail(StageFetch, "response body is not a JSON object")
	}

	if version, ok := jsonString(probe, "version"); !ok || version != "1" {
		return nil, fail(StageFetch, "unsupported bundle version: %s", jsonFieldOrUndefined(probe, "version"))
	}
	kind, ok := jsonString(probe, "kind")
	if !ok || !EvidenceKind(kind).valid() {
		return nil, fail(StageFetch, "unsupported bundle kind: %s", jsonFieldOrUndefined(probe, "kind"))
	}
	hostname, ok := jsonString(probe, "hostname")
	if !ok || hostname == "" {
		return nil, fail(StageFetch, "bundle is missing hostname")
	}
	if hostname != expectedHostname {
		return nil, fail(StageFetch, "bundle hostname %q does not match request hostname", hostname)
	}
	issuedAt, ok := jsonString(probe, "issuedAt")
	if !ok {
		return nil, fail(StageFetch, "bundle is missing issuedAt")
	}
	certFingerprint, ok := jsonString(probe, "certFingerprint")
	if !ok || !strings.HasPrefix(certFingerprint, FingerprintPrefix) {
		return nil, fail(StageFetch, "bundle certFingerprint is malformed")
	}
	jws, ok := jsonString(probe, "jws")
	if !ok || jws == "" {
		return nil, fail(StageFetch, "bundle is missing jws")
	}
	var certChain []string
	if raw, present := probe["certChain"]; !present || json.Unmarshal(raw, &certChain) != nil || len(certChain) == 0 {
		return nil, fail(StageFetch, "bundle certChain is missing or malformed")
	}
	for _, entry := range certChain {
		if entry == "" {
			return nil, fail(StageFetch, "bundle certChain is missing or malformed")
		}
	}

	bundle := &Bundle{
		Version:         "1",
		Kind:            EvidenceKind(kind),
		Hostname:        hostname,
		IssuedAt:        issuedAt,
		CertFingerprint: certFingerprint,
		JWS:             jws,
		CertChain:       certChain,
	}

	if raw, present := probe["rootCaTeeQuote"]; present && string(raw) != "null" {
		if err := json.Unmarshal(raw, &bundle.RootCaTeeQuote); err != nil {
			return nil, fail(StageFetch, "bundle rootCaTeeQuote is malformed")
		}
	}
	if raw, present := probe["tlsLeaf"]; present {
		if json.Unmarshal(raw, &bundle.TLSLeaf) != nil || bundle.TLSLeaf == "" {
			return nil, fail(StageFetch, "bundle tlsLeaf is malformed")
		}
	}
	return bundle, Result{}
}

func matchTrustedRoot(rootFingerprint string, trustedRoots []TrustedRoot) (*MatchedRoot, Result) {
	for _, root := range trustedRoots {
		fingerprint, err := RootFingerprintFromPEM(root.PEM)
		if err != nil {
			return nil, fail(StageUntrustedRoot, "failed to parse trusted root %q: %v", root.Name, err)
		}
		if FingerprintsEqual(fingerprint, rootFingerprint) {
			return &MatchedRoot{Name: root.Name, Fingerprint: fingerprint}, Result{}
		}
	}
	return nil, fail(StageUntrustedRoot, "%s not in trusted store", rootFingerprint)
}

func verifyPayload(bundle *Bundle, leaf *certparse.Certificate, expectedHostname string) (Payload, Result) {
	payload, err := verifyJWS(bundle.JWS, leaf)
	if err != nil {
		return nil, fail(StageJWS, "%v", err)
	}
	base := payload.Base()
	if base.Kind != bundle.Kind {
		return nil, fail(StageJWS, "payload kind %q does not match bundle kind %q", base.Kind, bundle.Kind)
	}
	if base.Hostname != expectedHostname {
		return nil, fail(StageJWS, "payload hostname %q does not match request hostname", base.Hostname)
	}
	return payload, Result{}
}

// checkFreshness rejects bundles older than maxAge, and bundles dated further
// into the future than AllowedClockSkew. A stale bundle is a replay risk: it
// proves what the endpoint looked like once, not what it looks like now.
// Reported at the jws stage, since issuedAt is a signed payload claim.
func checkFreshness(issuedAt string, maxAge time.Duration, now time.Time) (Result, bool) {
	if maxAge <= 0 {
		return Result{}, true
	}
	issued, err := time.Parse(time.RFC3339, issuedAt)
	if err != nil {
		return fail(StageJWS, "payload.issuedAt %q is not a parseable timestamp", issuedAt), false
	}
	age := now.Sub(issued)
	if age > maxAge {
		return fail(StageJWS, "bundle age %dms exceeds maxBundleAge=%dms",
			age.Milliseconds(), maxAge.Milliseconds()), false
	}
	if age < -AllowedClockSkew {
		return fail(StageJWS, "payload.issuedAt is %dms in the future, beyond allowed skew",
			(-age).Milliseconds()), false
	}
	return Result{}, true
}

// checkChannelBinding ties the signed certFingerprint to a concrete TLS
// certificate. The observed fingerprint always wins when the caller has one;
// hashing the bundle's own tlsLeaf is the weaker fallback for verifiers without
// channel access, and the gatekeeper never takes it in production.
func checkChannelBinding(payloadFingerprint, observedFingerprint, tlsLeafPEM string) (ChannelBinding, Result) {
	if observedFingerprint != "" {
		if !FingerprintsEqual(payloadFingerprint, observedFingerprint) {
			return "", fail(StageTLSFingerprint,
				"payload certFingerprint %s does not match observed %s", payloadFingerprint, observedFingerprint)
		}
		return BindingObserved, Result{}
	}

	if tlsLeafPEM != "" {
		cert, err := certparse.ParsePEM(tlsLeafPEM)
		if err != nil {
			return "", fail(StageTLSFingerprint, "failed to parse bundle.tlsLeaf: %v", err)
		}
		derived := SHA256Fingerprint(cert.Raw)
		if !FingerprintsEqual(payloadFingerprint, derived) {
			return "", fail(StageTLSFingerprint,
				"payload certFingerprint %s does not match bundle.tlsLeaf fingerprint %s", payloadFingerprint, derived)
		}
		return BindingProducerAsserted, Result{}
	}

	return "", fail(StageTLSFingerprint, "no observed fingerprint and no tlsLeaf in bundle")
}

// jsonString reads a field that must be a JSON string.
func jsonString(probe map[string]json.RawMessage, key string) (string, bool) {
	raw, present := probe[key]
	if !present {
		return "", false
	}
	var value string
	if err := json.Unmarshal(raw, &value); err != nil {
		return "", false
	}
	return value, true
}

// jsonFieldOrUndefined renders a field for an error message the way the
// TypeScript verifier does: the bare string when it is one, its JSON otherwise.
func jsonFieldOrUndefined(probe map[string]json.RawMessage, key string) string {
	raw, present := probe[key]
	if !present {
		return "undefined"
	}
	if value, ok := jsonString(probe, key); ok {
		return value
	}
	return string(raw)
}

File: libs/attestation/src/verify.ts
/**
 * The verification pipeline: fetch → cert chain → trusted root → JWS → freshness →
 * channel binding.
 *
 * Ported from Super-Protocol/swarm-cloud `libs/swarm-attestation/src/verify.ts`
 * (BSL-1.1) with permission; see the repository NOTICE. The Go verifier in
 * `apps/gatekeeper/pkg/attestation` implements the same stages and is held to the
 * same conformance vectors (`@confidential-router/attestation-fixtures`).
 */
import { X509Certificate } from '@peculiar/x509';
import { buildCacheKey } from './cache.js';
import { CertChainError, rootFingerprintFromPem, validateChain } from './cert-chain.js';
import { fail } from './errors.js';
import { fingerprintsEqual, isFingerprint, sha256Fingerprint } from './fingerprint.js';
import { JwsError, verifyJws } from './jws.js';
import type {
  AttestationBundle,
  ChannelBinding,
  EvidenceKind,
  EvidencePayload,
  TrustedRoot,
  VerifyParams,
  VerifyResult,
} from './types.js';

export const EVIDENCE_PATH = '/.well-known/swarm-evidence';

const KIND_VALUES = new Set<EvidenceKind>([
  'DeploymentEvidence',
  'ControlPlaneEvidence',
  'KubernetesControlPlaneEvidence',
]);
// Tolerance for benign clock skew between producer and verifier when `maxBundleAge` is
// set. A bundle dated up to this much in the future is still accepted; anything beyond
// is treated as a freshness violation.
export const ALLOWED_CLOCK_SKEW_MS = 60_000;

type StageResult<T> = ({ ok: true } & T) | { ok: false; err: VerifyResult };

export async function verifyHostname(params: VerifyParams): Promise<VerifyResult> {
  const { hostname, observedTlsFingerprint, trustedRoots, cache } = params;

  if (typeof hostname !== 'string' || hostname.length === 0) {
    return fail('fetch', 'hostname must be a non-empty string');
  }
  if (observedTlsFingerprint !== undefined && !isFingerprint(observedTlsFingerprint)) {
    return fail('tls-fingerprint', 'observedTlsFingerprint must match sha256/<base64url>');
  }
  if (!Array.isArray(trustedRoots)) {
    return fail('untrusted-root', 'trustedRoots must be an array (may be empty)');
  }

  const baseCacheKey = cache ? await buildCacheKey(hostname, observedTlsFingerprint ?? null, trustedRoots) : null;
  // Different age policies must not share cached results — a permissive caller's "ok"
  // could otherwise satisfy a stricter caller for a bundle that has since exceeded the
  // stricter window.
  const cacheKey = baseCacheKey ? `${baseCacheKey}|${params.maxBundleAge ?? ''}` : null;
  if (cache && cacheKey) {
    const hit = cache.get(cacheKey);
    if (hit) return hit;
  }

  const result = await runPipeline(params);

  if (cache && cacheKey && result.ok) {
    cache.set(cacheKey, result);
  }
  return result;
}

async function runPipeline(params: VerifyParams): Promise<VerifyResult> {
  const fetched = await fetchBundle(params);
  if (!fetched.ok) return fetched.err;
  const bundle = fetched.bundle;

  const chain = await runCertChainStage(bundle, params.now);
  if (!chain.ok) return chain.err;

  const trust = await runTrustedRootStage(chain.rootFingerprint, params.trustedRoots);
  if (!trust.ok) return trust.err;

  const jws = await runJwsStage(bundle, chain.leaf, params.hostname);
  if (!jws.ok) return jws.err;

  const freshness = runFreshnessStage(jws.payload.issuedAt, params.maxBundleAge, params.now);
  if (!freshness.ok) return freshness.err;

  const tls = await runTlsFingerprintStage(jws.payload.certFingerprint, params.observedTlsFingerprint, bundle.tlsLeaf);
  if (!tls.ok) return tls.err;

  return {
    ok: true,
    kind: jws.payload.kind,
    payload: jws.payload,
    matchedRoot: trust.matchedRoot,
    rootCaTeeQuote: bundle.rootCaTeeQuote,
    channelBinding: tls.channelBinding,
  };
}

async function fetchBundle(params: VerifyParams): Promise<StageResult<{ bundle: AttestationBundle }>> {
  const url = `https://${params.hostname}${EVIDENCE_PATH}`;
  const fetcher = params.fetcher ?? globalThis.fetch;
  if (typeof fetcher !== 'function') {
    return {
      ok: false,
      err: fail('fetch', 'no fetcher available; pass params.fetcher or run with global fetch'),
    };
  }

  let response: Response;
  try {
    response = await fetcher(url, {
      method: 'GET',
      headers: { accept: 'application/json' },
    });
  } catch (err) {
    return { ok: false, err: fail('fetch', `request failed: ${(err as Error).message}`) };
  }

  if (!response.ok) {
    return {
      ok: false,
      err: fail('fetch', `unexpected status ${response.status} from ${url}`),
    };
  }

  let body: unknown;
  try {
    body = await response.json();
  } catch (err) {
    return {
      ok: false,
      err: fail('fetch', `failed to parse response body as JSON: ${(err as Error).message}`),
    };
  }

  return validateBundleShape(body, params.hostname);
}

function validateBundleShape(raw: unknown, expectedHostname: string): StageResult<{ bundle: AttestationBundle }> {
  if (raw === null || typeof raw !== 'object') {
    return { ok: false, err: fail('fetch', 'response body is not a JSON object') };
  }
  const v = raw as Record<string, unknown>;
  if (v.version !== '1') {
    return { ok: false, err: fail('fetch', `unsupported bundle version: ${String(v.version)}`) };
  }
  if (typeof v.kind !== 'string' || !KIND_VALUES.has(v.kind as EvidenceKind)) {
    return { ok: false, err: fail('fetch', `unsupported bundle kind: ${String(v.kind)}`) };
  }
  if (typeof v.hostname !== 'string' || v.hostname.length === 0) {
    return { ok: false, err: fail('fetch', 'bundle is missing hostname') };
  }
  if (v.hostname !== expectedHostname) {
    return {
      ok: false,
      err: fail('fetch', `bundle hostname "${v.hostname}" does not match request hostname`),
    };
  }
  if (typeof v.issuedAt !== 'string') {
    return { ok: false, err: fail('fetch', 'bundle is missing issuedAt') };
  }
  if (typeof v.certFingerprint !== 'string' || !v.certFingerprint.startsWith('sha256/')) {
    return { ok: false, err: fail('fetch', 'bundle certFingerprint is malformed') };
  }
  if (typeof v.jws !== 'string' || v.jws.length === 0) {
    return { ok: false, err: fail('fetch', 'bundle is missing jws') };
  }
  if (
    !Array.isArray(v.certChain) ||
    v.certChain.length === 0 ||
    !v.certChain.every((p) => typeof p === 'string' && p.length > 0)
  ) {
    return { ok: false, err: fail('fetch', 'bundle certChain is missing or malformed') };
  }
  if (v.rootCaTeeQuote !== undefined && (typeof v.rootCaTeeQuote !== 'object' || v.rootCaTeeQuote === null)) {
    return { ok: false, err: fail('fetch', 'bundle rootCaTeeQuote is malformed') };
  }
  if (v.tlsLeaf !== undefined && (typeof v.tlsLeaf !== 'string' || v.tlsLeaf.length === 0)) {
    return { ok: false, err: fail('fetch', 'bundle tlsLeaf is malformed') };
  }
  return {
    ok: true,
    bundle: {
      version: '1',
      kind: v.kind as EvidenceKind,
      hostname: v.hostname,
      issuedAt: v.issuedAt,
      certFingerprint: v.certFingerprint,
      jws: v.jws,
      certChain: v.certChain as string[],
      rootCaTeeQuote: v.rootCaTeeQuote as AttestationBundle['rootCaTeeQuote'],
      tlsLeaf: v.tlsLeaf as string | undefined,
    },
  };
}

async function runCertChainStage(
  bundle: AttestationBundle,
  now: Date | undefined,
): Promise<StageResult<{ rootFingerprint: string; leaf: X509Certificate }>> {
  try {
    const parsed = await validateChain(bundle.certChain, { now });
    return { ok: true, rootFingerprint: parsed.rootFingerprint, leaf: parsed.leaf };
  } catch (err) {
    if (err instanceof CertChainError) {
      return { ok: false, err: fail('cert-chain', err.message) };
    }
    return { ok: false, err: fail('cert-chain', `unexpected error: ${(err as Error).message}`) };
  }
}

async function runTrustedRootStage(
  rootFingerprint: string,
  trustedRoots: TrustedRoot[],
): Promise<StageResult<{ matchedRoot: { name: string; fingerprint: string } }>> {
  for (const root of trustedRoots) {
    let fp: string;
    try {
      fp = await rootFingerprintFromPem(root.pem);
    } catch (err) {
      return {
        ok: false,
        err: fail('untrusted-root', `failed to parse trusted root "${root.name}": ${(err as Error).message}`),
      };
    }
    if (fingerprintsEqual(fp, rootFingerprint)) {
      return { ok: true, matchedRoot: { name: root.name, fingerprint: fp } };
    }
  }
  return {
    ok: false,
    err: fail('untrusted-root', `${rootFingerprint} not in trusted store`),
  };
}

async function runJwsStage(
  bundle: AttestationBundle,
  leaf: X509Certificate,
  expectedHostname: string,
): Promise<StageResult<{ payload: EvidencePayload }>> {
  try {
    const payload = await verifyJws(bundle.jws, leaf);
    if (payload.kind !== bundle.kind) {
      return {
        ok: false,
        err: fail('jws', `payload kind "${payload.kind}" does not match bundle kind "${bundle.kind}"`),
      };
    }
    if (payload.hostname !== expectedHostname) {
      return {
        ok: false,
        err: fail('jws', `payload hostname "${payload.hostname}" does not match request hostname`),
      };
    }
    return { ok: true, payload };
  } catch (err) {
    if (err instanceof JwsError) {
      return { ok: false, err: fail('jws', err.message) };
    }
    return { ok: false, err: fail('jws', `unexpected error: ${(err as Error).message}`) };
  }
}

function runFreshnessStage(
  issuedAt: string,
  maxBundleAge: number | undefined,
  now: Date | undefined,
): StageResult<Record<never, never>> {
  if (typeof maxBundleAge !== 'number') return { ok: true };
  const issued = Date.parse(issuedAt);
  if (!Number.isFinite(issued)) {
    return { ok: false, err: fail('jws', `payload.issuedAt "${issuedAt}" is not a parseable timestamp`) };
  }
  const age = (now ?? new Date()).getTime() - issued;
  if (age > maxBundleAge) {
    return { ok: false, err: fail('jws', `bundle age ${age}ms exceeds maxBundleAge=${maxBundleAge}ms`) };
  }
  if (age < -ALLOWED_CLOCK_SKEW_MS) {
    return { ok: false, err: fail('jws', `payload.issuedAt is ${-age}ms in the future, beyond allowed skew`) };
  }
  return { ok: true };
}

async function runTlsFingerprintStage(
  payloadFingerprint: string,
  observedFingerprint: string | undefined,
  tlsLeafPem: string | undefined,
): Promise<StageResult<{ channelBinding: ChannelBinding }>> {
  // Live channel binding is preferred whenever the caller can produce it — it is the
  // only mode the gatekeeper uses. Producer-asserted binding is the fallback for
  // environments without channel access.
  if (observedFingerprint !== undefined) {
    if (!fingerprintsEqual(payloadFingerprint, observedFingerprint)) {
      return {
        ok: false,
        err: fail(
          'tls-fingerprint',
          `payload certFingerprint ${payloadFingerprint} does not match observed ${observedFingerprint}`,
        ),
      };
    }
    return { ok: true, channelBinding: 'observed' };
  }

  if (tlsLeafPem !== undefined) {
    let derivedFingerprint: string;
    try {
      const cert = new X509Certificate(tlsLeafPem);
      derivedFingerprint = await sha256Fingerprint(new Uint8Array(cert.rawData));
    } catch (err) {
      return {
        ok: false,
        err: fail('tls-fingerprint', `failed to parse bundle.tlsLeaf: ${(err as Error).message}`),
      };
    }
    if (!fingerprintsEqual(payloadFingerprint, derivedFingerprint)) {
      return {
        ok: false,
        err: fail(
          'tls-fingerprint',
          `payload certFingerprint ${payloadFingerprint} does not match bundle.tlsLeaf fingerprint ${derivedFingerprint}`,
        ),
      };
    }
    return { ok: true, channelBinding: 'producer-asserted' };
  }

  return {
    ok: false,
    err: fail('tls-fingerprint', 'no observed fingerprint and no tlsLeaf in bundle'),
  };
}
